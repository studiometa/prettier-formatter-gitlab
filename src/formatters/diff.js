import chalk from 'chalk';
import { diffStringsUnified } from 'jest-diff';
import { getPrettierFileInfos } from '../utils/get-prettier-file-infos.js';

/**
 * @typedef {import('../types.d.ts').FileInfo} FileInfo
 */

/**
 * Format Prettier formatting errors in diff format.
 * @param {Partial<FileInfo>} fileInfo A file Prettier informations.
 * @returns {string}
 */
function formatDiff({ filename, input, output, error } = {}) {
  let diff = `----------------------------------------------------
${filename}
----------------------------------------------------
`;

  if (error) {
    diff += String(error);
  } else {
    diff += diffStringsUnified(input || '', output || '', {
      aColor: chalk.red,
      bColor: chalk.green,
      omitAnnotationLines: true,
      contextLines: 2,
      expand: false,
    });
  }

  diff += '\n';

  return diff;
}

/**
 * Diff formatter.
 * @param   {string[]} files
 * @returns {Promise<void>}
 */
export async function diff(files) {
  const infos = await Promise.all(files.map((file) => getPrettierFileInfos(file)));
  return infos.map((info) => formatDiff(info));
}
