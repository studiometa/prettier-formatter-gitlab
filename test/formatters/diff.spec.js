import { describe, it, expect, spyOn } from 'bun:test';
import { resolve } from 'node:path';
import { diff } from '../../src/formatters/diff.js';

describe('The diff formatter', () => {
  it('should return diff formatted reports about the given files', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/dirty.js');
    const results = await diff([filepath]);
    expect(results.join('\n')).toMatchSnapshot();
  });

  it('should handle files with syntax errors', async () => {
    const filepath = resolve(import.meta.dirname, '../__stubs__/syntax-error.js');
    const results = await diff([filepath]);
    expect(results.join('\n')).toMatchSnapshot();
  });
});
