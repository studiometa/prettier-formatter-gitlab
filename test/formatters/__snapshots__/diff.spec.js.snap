// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`The diff formatter should return diff formatted reports about the given files 1`] = `
"----------------------------------------------------
test/__stubs__/dirty.js
----------------------------------------------------
\x1B[33m@@ -1,18 +1,7 @@\x1B[39m
\x1B[2m  /* eslint-disable */\x1B[22m

\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m- function foo(\x1B[7m \x1B[27marg\x1B[7m  \x1B[27m) {\x1B[39m
\x1B[32m+ function foo(arg) {\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[2m    const data = arg.split(',');\x1B[22m

\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[31m-\x1B[39m
\x1B[2m    return data;\x1B[22m
\x1B[2m  }\x1B[22m
"
`;

exports[`The diff formatter should handle files with syntax errors 1`] = `
"----------------------------------------------------
test/__stubs__/syntax-error.js
----------------------------------------------------
SyntaxError: Unexpected token, expected "(" (3:13)
\x1B[0m \x1B[90m 1 |\x1B[39m \x1B[90m/* eslint-disable */\x1B[39m
 \x1B[90m 2 |\x1B[39m
\x1B[31m\x1B[1m>\x1B[22m\x1B[39m\x1B[90m 3 |\x1B[39m \x1B[36mfunction\x1B[39m foo\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m\x1B[33m#\x1B[39m( arg  ) {
 \x1B[90m   |\x1B[39m             \x1B[31m\x1B[1m^\x1B[22m\x1B[39m
 \x1B[90m 4 |\x1B[39m   \x1B[36mreturn\x1B[39m arg\x1B[33m;\x1B[39m
 \x1B[90m 5 |\x1B[39m }
 \x1B[90m 6 |\x1B[39m\x1B[0m
"
`;
